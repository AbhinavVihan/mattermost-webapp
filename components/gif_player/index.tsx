// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, {useCallback, useEffect, useState} from 'react';
import classNames from 'classnames';

import './gif_player.scss';

const preload = (src: string, callback: (img: HTMLImageElement) => void) => {
    const img = new Image();
    if (typeof callback === 'function') {
        img.onload = () => callback(img);

        // img.setAttribute('crossOrigin', 'anonymous');
    }
    img.src = src;
};

const firstGifFrameUrl = (img: HTMLImageElement) => {
    const canvas = document.createElement('canvas');
    if (typeof canvas.getContext !== 'function') {
        return null;
    }
    canvas.width = img.width;
    canvas.height = img.height;
    const ctx = canvas.getContext('2d');
    ctx?.drawImage(img, 0, 0);
    console.log('props canvas', img, canvas.toDataURL());
    return canvas.toDataURL();
};

type Props = {
    gif: string;
    still: string;
    playing: boolean;
    toggle: () => void;
    propagation?: boolean;
    playButton?: boolean;
}

function GifPlayer({gif, still, playing, toggle, propagation, playButton}: Props) {
    return (
        <div
            className={classNames('gif_player', {playing})}
            onClick={propagation ? undefined : toggle}
        >
            <div className={playButton ? 'play_button' : ''}/>
            <img
                src={playing ? (gif || still) : (still || gif)}
            />
        </div>
    );
}

type GifCTRProps = {
    gif: string;
    still?: string;
    autoPlay: boolean;
    propagation?: boolean;
    playButton?: boolean;
}

function GifPlayerContainer(props: GifCTRProps) {
    const [playing, setPlaying] = useState(props.autoPlay);
    const [actualStill, setActualStill] = useState<string | null>(null);

    useEffect(() => {
        preload(props.gif, (img) => {
            console.log('props gif', props.gif);
            const actualStill = firstGifFrameUrl(img);
            if (actualStill) {
                setActualStill(actualStill);
            }
        },
        );
    }, [props.gif, props.still]);

    const toggle = useCallback(() => {
        setPlaying(!playing);
    }, [playing]);

    return (
        <GifPlayer
            gif={props.gif}
            still={actualStill!}
            playing={playing}
            toggle={toggle}
            propagation={props.propagation}
            playButton={props.playButton}
        />
    );
}

export default GifPlayerContainer;
