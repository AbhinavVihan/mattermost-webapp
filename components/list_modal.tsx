// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Modal} from 'react-bootstrap';
import {FormattedMessage} from 'react-intl';

import LoadingScreen from 'components/loading_screen';
import {Group} from '@mattermost/types/groups';

export const DEFAULT_NUM_PER_PAGE = 50;

type Props = {
    titleText?: string;
    searchPlaceholderText?: string;
    titleBarButtonText?: string;
    numPerPage?: number;
    show?: boolean;
    titleBarButtonOnClick?: () => void;
    loadItems: (page: number, searchTerm: string) => Promise<{
        items: Group[];
        totalCount: number;
    }>;
    onHide?: () => void;
    renderRow: (item: Group, listModal: ListModal) => JSX.Element;
}

type State = {
    show: boolean;
    page: number;
    items: Group[];
    totalCount: number;
    loading: boolean;
    searchTerm: string;
}

export default class ListModal extends React.PureComponent<Props, State> {
    static defaultProps = {
        show: true,
    };
    numPerPage: number;

    constructor(props: Props) {
        super(props);

        this.numPerPage = props.numPerPage || DEFAULT_NUM_PER_PAGE;

        this.state = {
            show: true,
            page: 0,
            items: [],
            totalCount: 0,
            loading: true,
            searchTerm: '',
        };
    }

    async componentDidMount() {
        const {totalCount, items} = await this.props.loadItems(0, '');
        this.setState({totalCount, items, loading: false}); // eslint-disable-line react/no-did-mount-set-state
    }

    handleHide = () => {
        this.setState({show: false});
    }

    handleExit = () => {
        if (this.props.onHide) {
            this.props.onHide();
        }
    }

    renderRows() {
        if (this.state.loading) {
            return (
                <div>
                    <LoadingScreen
                        position='absolute'
                        key='loading'
                    />
                </div>
            );
        }
        return this.state.items.map((item) => (
            this.props.renderRow(item, this)
        ));
    }

    onNext = () => {
        const nextPage = this.state.page + 1;
        this.onPageChange(nextPage);
    }

    onPrev = () => {
        const prevPage = this.state.page - 1;
        this.onPageChange(prevPage);
    }

    onPageChange = async (page: number) => {
        this.setState({loading: true});
        const result = await this.props.loadItems(page, this.state.searchTerm);
        this.setState({page, items: result.items, loading: false});
    }

    onSearchInput = async (event: React.ChangeEvent<HTMLInputElement>) => {
        const {target} = event;
        const searchTerm = target.value;
        this.setState({loading: true, searchTerm});
        const result = await this.props.loadItems(0, searchTerm);
        const {items, totalCount} = result;
        this.setState({loading: false, items, totalCount});
    }

    paginationRange() {
        let startCount = (this.state.page * this.numPerPage) + 1;
        const endCount = (startCount + this.state.items.length) - 1;
        if (endCount === 0) {
            startCount = 0;
        }
        return {startCount, endCount};
    }

    render() {
        if (!this.props.show) {
            return null;
        }
        const {endCount, startCount} = this.paginationRange();
        return (
            <div>
                <Modal
                    dialogClassName='a11y__modal more-modal more-modal--action'
                    show={this.state.show}
                    onHide={this.handleHide}
                    onExited={this.handleExit}
                >
                    <Modal.Header closeButton={true}>
                        <Modal.Title componentClass='h1'>
                            <span className='name'>{this.props.titleText}</span>
                        </Modal.Title>
                        {this.props.titleBarButtonText && this.props.titleBarButtonOnClick &&
                            <a
                                className='btn btn-md btn-primary'
                                href='#'
                                onClick={this.props.titleBarButtonOnClick}
                            >
                                {this.props.titleBarButtonText}
                            </a>}
                    </Modal.Header>
                    <Modal.Body>
                        <div className='filtered-user-list'>
                            <div className='filter-row'>
                                <div className='col-xs-12'>
                                    <label
                                        className='hidden-label'
                                        htmlFor='searchUsersInput'
                                    >
                                        {this.props.searchPlaceholderText}
                                    </label>
                                    <input
                                        id='searchUsersInput'
                                        className='form-control filter-textbox'
                                        placeholder={this.props.searchPlaceholderText}
                                        onChange={this.onSearchInput}
                                    />
                                </div>
                                <div className='col-sm-12'>
                                    <span className='member-count pull-left'>
                                        <FormattedMessage
                                            id='list_modal.paginatorCount'
                                            defaultMessage='{startCount, number} - {endCount, number} of {total, number} total'
                                            values={{
                                                startCount,
                                                endCount,
                                                total: this.state.totalCount,
                                            }}
                                        />
                                    </span>
                                </div>
                            </div>
                            <div className='more-modal__list'>
                                <div>
                                    {this.renderRows()}
                                </div>
                            </div>
                            <div className='filter-controls'>
                                {this.state.page > 0 &&
                                <button
                                    onClick={this.onPrev}
                                    className='btn btn-link filter-control filter-control__prev'
                                >
                                    <FormattedMessage
                                        id='filtered_user_list.prev'
                                        defaultMessage='Previous'
                                    />
                                </button>}
                                {this.props.numPerPage && (this.state.items.length >= this.props.numPerPage) && endCount !== this.state.totalCount &&
                                <button
                                    onClick={this.onNext}
                                    className='btn btn-link filter-control filter-control__next'
                                >
                                    <FormattedMessage
                                        id='filtered_user_list.next'
                                        defaultMessage='Next'
                                    />
                                </button>}
                            </div>
                        </div>
                    </Modal.Body>
                </Modal>
            </div >
        );
    }
}
