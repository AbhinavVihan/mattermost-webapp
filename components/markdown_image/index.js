// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import {openModal} from 'actions/views/modals';

import {get} from 'mattermost-redux/selectors/entities/preferences';
import {Preferences} from 'utils/constants';

import MarkdownImage from './markdown_image';

function mapStateToProps(state) {
    return {
        autoplayGifAndEmojis: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.AUTOPLAY_GIF_AND_EMOJI, Preferences.LINK_PREVIEW_DISPLAY_DEFAULT),
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            openModal,
        }, dispatch),
    };
}

const connector = connect(mapStateToProps, mapDispatchToProps);

export default connector(MarkdownImage);
