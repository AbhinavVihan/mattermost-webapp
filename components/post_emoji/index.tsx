// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';

import {getEmojiImageUrl} from 'mattermost-redux/utils/emoji_utils';

import {getEmojiMap} from 'selectors/emojis';

import {GlobalState} from 'types/store';

import {get} from 'mattermost-redux/selectors/entities/preferences';
import {Preferences} from 'utils/constants';

import PostEmoji from './post_emoji';

type Props = {
    name: string;
};

function mapStateToProps(state: GlobalState, ownProps: Props) {
    const emojiMap = getEmojiMap(state);
    const emoji = emojiMap.get(ownProps.name);

    return {
        imageUrl: emoji ? getEmojiImageUrl(emoji) : '',
        autoplayGifAndEmojis: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.AUTOPLAY_GIF_AND_EMOJI, Preferences.LINK_PREVIEW_DISPLAY_DEFAULT),
    };
}

export default connect(mapStateToProps)(PostEmoji);
