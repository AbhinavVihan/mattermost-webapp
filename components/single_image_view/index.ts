// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect, ConnectedProps} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';

import {GenericAction} from 'mattermost-redux/types/actions';
import {getConfig} from 'mattermost-redux/selectors/entities/general';

import {GlobalState} from 'types/store';

import {toggleEmbedVisibility} from 'actions/post_actions';
import {openModal} from 'actions/views/modals';
import {getFilePublicLink} from 'mattermost-redux/actions/files';

import {getIsRhsOpen} from 'selectors/rhs';

import SingleImageView from 'components/single_image_view/single_image_view';
import {get} from 'mattermost-redux/selectors/entities/preferences';
import {Preferences} from 'utils/constants';

function mapStateToProps(state: GlobalState) {
    const isRhsOpen = getIsRhsOpen(state);
    const config = getConfig(state);

    return {
        isRhsOpen,
        enablePublicLink: config.EnablePublicLink === 'true',
        autoplayGifAndEmojis: get(state, Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.AUTOPLAY_GIF_AND_EMOJI, Preferences.LINK_PREVIEW_DISPLAY_DEFAULT),
    };
}

function mapDispatchToProps(dispatch: Dispatch<GenericAction>) {
    return {
        actions: bindActionCreators({
            toggleEmbedVisibility,
            openModal,
            getFilePublicLink,
        }, dispatch),
    };
}

const connector = connect(mapStateToProps, mapDispatchToProps);

export type PropsFromRedux = ConnectedProps<typeof connector>;

export default connector(SingleImageView);
