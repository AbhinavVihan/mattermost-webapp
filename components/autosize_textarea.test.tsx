// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {AutosizeTextarea} from 'components/autosize_textarea';
import {renderWithIntl} from 'tests/react_testing_utils';

describe('components/AutosizeTextarea', () => {
    test('should match snapshot, init', () => {
        const wrapper = renderWithIntl(
            <AutosizeTextarea/>,
        );

        expect(wrapper).toMatchSnapshot();
    });
});
