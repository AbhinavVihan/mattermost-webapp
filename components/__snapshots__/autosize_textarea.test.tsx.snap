// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AutosizeTextarea should match snapshot, init 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          data-testid="autosize_textarea_placeholder"
          id="autosize_textarea_placeholder"
          style="overflow: hidden; text-overflow: ellipsis; opacity: 0.5; pointer-events: none; position: absolute; white-space: nowrap; background: none; border-color: transparent;"
        />
        <textarea
          aria-label=""
          data-testid="autosize_textarea"
          dir="auto"
          height="0"
          id="autosize_textarea"
          role="textbox"
          rows="1"
        />
        <div
          style="height: 0px; overflow: hidden;"
        >
          <textarea
            aria-hidden="true"
            dir="auto"
            disabled=""
            id="autosize_textarea-reference"
            rows="1"
            style="height: auto; width: 100%;"
          />
          <div
            id="autosize_textarea-measuring"
            style="width: auto; display: inline-block;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        data-testid="autosize_textarea_placeholder"
        id="autosize_textarea_placeholder"
        style="overflow: hidden; text-overflow: ellipsis; opacity: 0.5; pointer-events: none; position: absolute; white-space: nowrap; background: none; border-color: transparent;"
      />
      <textarea
        aria-label=""
        data-testid="autosize_textarea"
        dir="auto"
        height="0"
        id="autosize_textarea"
        role="textbox"
        rows="1"
      />
      <div
        style="height: 0px; overflow: hidden;"
      >
        <textarea
          aria-hidden="true"
          dir="auto"
          disabled=""
          id="autosize_textarea-reference"
          rows="1"
          style="height: auto; width: 100%;"
        />
        <div
          id="autosize_textarea-measuring"
          style="width: auto; display: inline-block;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
